onchain/cairo/afk/src/dao/dao_aa.cairo

```diff
@@ -150,73 +167,160 @@ pub mod DaoAA {
         // TODO
         // Check if ERC20 minimal balance to create a proposal is needed, if yes check the  balance
         // Add TX Calldata for this proposal
-        fn create_proposal(ref self: ContractState, proposal: Proposal) {
-            let caller = get_caller_address();
-            let proposal_id = self.total_proposal.read();
-            self.total_proposal.write(proposal_id + 1);
-            self.proposals.entry(proposal_id).write(proposal);
+        fn create_proposal(
+            ref self: ContractState, proposal_params: ProposalParams, calldata: Array<felt252>,
+        ) -> u256 {
+            assert(calldata.len() > 0, 'EMPTY CALLDATA');
+            let owner = get_caller_address();
+            let minimal_balance = self.minimal_balance_create_proposal.read();
+
+            if minimal_balance > 0 {
+                let vote_token_dispatcher = IERC20Dispatcher {
+                    contract_address: self.token_contract_address.read()
+                };
+                assert(
+                    vote_token_dispatcher.balance_of(owner) > minimal_balance,
+                    'INSUFFICIENT CREATION FUNDS'
+                );
+            }
 
-            let vote_state = VoteState {
-                votes_by_proposal: Map::new(), user_votes: Map::new(), has_voted: Map::new(),
+            let id = self.total_proposal.read() + 1;
+            let created_at = starknet::get_block_timestamp();
+            let end_at = starknet::get_block_timestamp() + SET_PROPOSAL_DURATION_IN_SECONDS;
+
+            let proposal = Proposal {
+                id,
+                created_at,
+                end_at,
+                is_whitelisted: false,
+                proposal_params,
+                proposal_status: Default::default(),
+                proposal_result: Default::default(),
+                proposal_result_at: 0,
+                owner,
+                proposal_result_by: contract_address_const::<0x0>(),
+                is_executed: false,
+                is_canceled: false,
             };
-            self.proposal_by_user.entry(caller).write(proposal_id);
-            self.vote_state_by_proposal.entry(proposal_id).write(vote_state);
-        }
 
-        // TODO
-        // Check if ERC20 minimal balance is needed
-        // Check if ERC20 max balance is needed
-        // Check is_multi_vote_available_per_token_balance
-        // Finish the voting part
-        fn cast_vote_type(ref self: ContractState, proposal_id: u256, vote: UserVote) {
-            let caller = get_caller_address();
-            self.vote_by_proposal.entry(proposal_id).write(vote);
-            self.user_votes.entry(caller).write(proposal_id);
-            self.has_voted.entry(caller).write(true);
-            self.user_vote_type.entry(caller).write(vote);
+            self.proposals.entry(id).write(Option::Some(proposal));
+
+            let proposal_calldata = self.proposals_calldata.entry(id);
```

Comment:
What is that loop? THe user need to provided calldata to save, here you do a loop of others calldata existing?